Decision records/10 - We use dedicated node services for server-side-rendering React in Rails applications.md

```diff
@@ -0,0 +1,37 @@
+
+Since a JavaScript runtime in some capacity is required many solutions involve a separate node server. Our largest React applications are totally independent node apps which interact with Rails entirely as an API server, and handle their own authentication, metrics and API proxying concerns. On the other hand other small applications opt out of SSR entirely and simply render a client-side React application from a Rails view.
+
+Both of these solutions have problems, with the prior adding more complexity at the server layer and requiring devs to maintain a separate server built on totally different technologies, and the latter losing out on performance and accessibility gains from SSR.
+
+## Solution
+
+We maintain our own javascript library, [`@shopify/react-server`](https://github.com/Shopify/quilt/tree/master/packages/react-server), built from the ground up to support [`@shopify/react-html`](https://github.com/Shopify/quilt/tree/master/packages/react-html) consuming React applications. On the Rails side we provide [`insert-gem-name`](https://github.com/Shopify/sewing-kit) to proxy requests to the rendering service. By proxying requests to node in this way we allow both our React and Rails patterns to remain consistent, with Rails developers continuing to handle authentication, API requests, and other server-side concerns the same way they would in a conventional Rails application, and React developers being able to build applications the same way that they would in a totally standalone node app. We use Shopify Build, Cloud Platform and ServicesDB to provide a template for shipping the Rails+React app as a single deploy from the same repo, while still allowing the two services to be scaled and monitored independentally.
+
+Our solution is actually quite similar to Airbnb's [`HyperNova`](https://github.com/airbnb/hypernova), however that library has it's own opinionated runtime associated which precludes the use of arbitrary serialization, and instead focuses on serializing state for specific libraries, as well as server-provided props for the top-level component. In addition, Airbnb itself has moved away from [`HyperNova`](https://github.com/airbnb/hypernova) citing concerns with having to write process supervisor trees in application code for larger projects. Since our solution is built from the ground up around letting the application code define it's own serialization / deserialization needs, and make use of our service scaling infrastructure, we can avoid the problems they encountered.
```

Comment:
```suggestion
Our solution is actually quite similar to Airbnb's [`HyperNova`](https://github.com/airbnb/hypernova), however that library has its own opinionated runtime associated which precludes the use of arbitrary serialization, and instead focuses on serializing state for specific libraries, as well as server-provided props for the top-level component. In addition, Airbnb itself has moved away from [`HyperNova`](https://github.com/airbnb/hypernova) citing concerns with having to write process supervisor trees in application code for larger projects. Since our solution is built from the ground up around letting the application code define its own serialization / deserialization needs, and make use of our service scaling infrastructure, we can avoid the problems they encountered.
```